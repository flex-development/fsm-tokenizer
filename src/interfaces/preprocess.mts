/**
 * @file Interfaces - Preprocess
 * @module fsm-tokenizer/interfaces/Preprocess
 */

import type {
  Code,
  Encoding,
  FileLike,
  Value
} from '@flex-development/fsm-tokenizer'

/**
 * Turn `value` into character code chunks.
 *
 * @see {@linkcode Code}
 * @see {@linkcode Encoding}
 * @see {@linkcode FileLike}
 * @see {@linkcode Value}
 */
interface Preprocess {
  /**
   * Turn `value` into character code chunks.
   *
   * @see {@linkcode Code}
   * @see {@linkcode Encoding}
   * @see {@linkcode FileLike}
   * @see {@linkcode Value}
   *
   * @this {void}
   *
   * @param {FileLike | Value | null | undefined} value
   *  The value to preprocess
   * @param {Encoding | null | undefined} encoding
   *  The character encoding to use when `value`
   *  or its contents is {@linkcode Uint8Array}
   * @param {true} end
   *  Whether the end of stream has been reached
   * @return {Code[]}
   *  Character code chunks
   */
  (
    this: void,
    value: FileLike | Value | null | undefined,
    encoding: Encoding | null | undefined,
    end: true
  ): Code[]

  /**
   * Turn `value` into character code chunks.
   *
   * @see {@linkcode Code}
   * @see {@linkcode Encoding}
   * @see {@linkcode FileLike}
   * @see {@linkcode Value}
   *
   * @this {void}
   *
   * @param {FileLike | Value | null | undefined} value
   *  The value to preprocess
   * @param {Encoding | null | undefined} encoding
   *  The character encoding to use when `value`
   *  or its contents is {@linkcode Uint8Array}
   * @param {false | null | undefined} end
   *  Whether the end of stream has been reached
   * @return {NonNullable<Code>[]}
   *  Character code chunks
   */
  (
    this: void,
    value: FileLike | Value | null | undefined,
    encoding: Encoding | null | undefined,
    end: false | null | undefined
  ): NonNullable<Code>[]

  /**
   * Turn `value` into character code chunks.
   *
   * @see {@linkcode Code}
   * @see {@linkcode Encoding}
   * @see {@linkcode FileLike}
   * @see {@linkcode Value}
   *
   * @this {void}
   *
   * @param {FileLike | Value | null | undefined} value
   *  The value to preprocess
   * @param {Encoding | null | undefined} [encoding]
   *  The character encoding to use when `value`
   *  or its contents is {@linkcode Uint8Array}
   * @param {false | null | undefined} [end]
   *  Whether the end of stream has been reached
   * @return {NonNullable<Code>[]}
   *  Character code chunks
   */
  (
    this: void,
    value: FileLike | Value | null | undefined,
    encoding?: Encoding | null | undefined,
    end?: false | null | undefined
  ): NonNullable<Code>[]

  /**
   * Turn `value` into character code chunks.
   *
   * @see {@linkcode Code}
   * @see {@linkcode Encoding}
   * @see {@linkcode FileLike}
   * @see {@linkcode Value}
   *
   * @this {void}
   *
   * @param {FileLike | Value | null | undefined} value
   *  The value to preprocess
   * @param {Encoding | null | undefined} [encoding]
   *  The character encoding to use when `value`
   *  or its contents is {@linkcode Uint8Array}
   * @param {boolean | null | undefined} [end]
   *  Whether the end of stream has been reached
   * @return {Code[]}
   *  Character code chunks
   */
  (
    this: void,
    value: FileLike | Value | null | undefined,
    encoding?: Encoding | null | undefined,
    end?: boolean | null | undefined
  ): Code[]
}

export type { Preprocess as default }
