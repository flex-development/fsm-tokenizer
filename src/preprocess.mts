/**
 * @file preprocess
 * @module vfile-tokenizer/preprocess
 */

import codes from '#enums/codes'
import constants from '#enums/constants'
import type {
  Code,
  Column,
  Encoding,
  FileLike,
  Preprocess,
  PreprocessOptions,
  Value
} from '@flex-development/vfile-tokenizer'

/**
 * Create a preprocessor to turn a value into character code chunks.
 *
 * @see {@linkcode Preprocess}
 * @see {@linkcode PreprocessOptions}
 *
 * @this {void}
 *
 * @param {PreprocessOptions | null | undefined} [options]
 *  Configuration options
 * @return {Preprocess}
 *  Character code preprocessor
 */
function preprocess(
  this: void,
  options?: PreprocessOptions | null | undefined
): Preprocess {
  return Object.defineProperties(preprocessor.bind(options ?? {}), {
    name: { value: 'preprocess' }
  })

  /**
   * Turn `value` into character code chunks.
   *
   * @this {PreprocessOptions}
   *
   * @param {FileLike | Value | null | undefined} value
   *  The value to preprocess
   * @param {Encoding | null | undefined} [encoding]
   *  The character encoding to use when `value`
   *  or its contents is {@linkcode Uint8Array}
   * @param {boolean | null | undefined} [end]
   *  Whether the end of stream has been reached
   * @return {Code[]}
   *  Character code chunks
   */
  function preprocessor(
    this: PreprocessOptions,
    value: FileLike | Value | null | undefined,
    encoding?: Encoding | null | undefined,
    end?: boolean | null | undefined
  ): Code[] {
    /**
     * Character code chunks.
     *
     * @const {Code[]} chunks
     */
    const chunks: Code[] = []

    /**
     * Number of spaces a tab is equivalent to.
     *
     * @const {number} tabSize
     */
    const tabSize: number = this.tabSize ?? constants.tabSize

    if (value === null) {
      chunks.push(codes.break)
    } else if (value !== undefined) {
      if (typeof value === 'object' && 'value' in value) {
        value = value.value
      }

      if (typeof value !== 'string') {
        value = new TextDecoder(encoding ?? undefined).decode(value)
      }

      /**
       * Current column.
       *
       * @var {Column} column
       */
      let column: Column = 1

      /**
       * Index of current character code.
       *
       * @var {number} index
       */
      let index: number = 0

      while (index < value.length) {
        /**
         * Character code.
         *
         * @var {NonNullable<Code>} code
         */
        let code: NonNullable<Code> = value[index]!.codePointAt(0)!

        /**
         * Difference between next column and current column.
         *
         * @var {number} k
         */
        let k: number = 1

        switch (true) {
          case code === codes.cr:
            if (value[index + 1]?.codePointAt(0) === codes.lf) {
              chunks.push(codes.crlf)
              k++
            } else {
              chunks.push(codes.vcr)
            }

            column = 1
            break
          case code === codes.ht:
            /**
             * Next column.
             *
             * @const {number} n
             */
            const n: number = Math.ceil(column / tabSize) * tabSize

            chunks.push(codes.vht)
            while (column++ < n) chunks.push(codes.vs)

            break
          case code === codes.lf:
            chunks.push(codes.vlf)
            column = 1
            break
          default:
            chunks.push(code)
            column++
            break
        }

        index += k
      }
    }

    return end && chunks.push(codes.eof), chunks
  }
}

export default preprocess
