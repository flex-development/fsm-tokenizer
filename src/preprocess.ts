/**
 * @file preprocess
 * @module vfile-lexer/preprocess
 */

import { codes } from './enums'
import type { PreprocessOptions } from './interfaces'
import type {
  Code,
  Column,
  Encoding,
  FileLike,
  Preprocessor,
  Value
} from './types'

/**
 * Create a preprocessor to turn a value into character code chunks.
 *
 * @see {@linkcode PreprocessOptions}
 * @see {@linkcode Preprocessor}
 *
 * @param {PreprocessOptions | null | undefined} [options] - Configuration
 * @return {Preprocessor} Character code preprocessor
 */
function preprocess(
  this: void,
  options?: PreprocessOptions | null | undefined
): Preprocessor {
  const { tabSize = 2 } = options ?? {}
  return preprocessor

  /**
   * Turn `value` into character code chunks.
   *
   * @param {FileLike | Value | null | undefined} value - Value to preprocess
   * @param {Encoding | null | undefined} [encoding] - Character encoding to use
   * when value or its contents is {@linkcode Uint8Array}
   * @param {boolean | null | undefined} [end] - End of stream?
   * @return {Code[]} Character code chunks
   */
  function preprocessor(
    value: FileLike | Value | null | undefined,
    encoding?: Encoding | null | undefined,
    end?: boolean | null | undefined
  ): Code[] {
    /**
     * Character code chunks.
     *
     * @const {Code[]} chunks
     */
    const chunks: Code[] = []

    if (
      (typeof value === 'string' && value) ||
      (typeof value === 'object' && value)
    ) {
      value = typeof value === 'object' && 'value' in value
        ? value.value
        : value

      value = typeof value === 'string'
        ? value.toString()
        : new TextDecoder(encoding ?? undefined).decode(value)

      /**
       * Current column.
       *
       * @var {Column} column
       */
      let column: Column = 1

      /**
       * Index of current character code.
       *
       * @var {number} index
       */
      let index: number = 0

      while (index < value.length) {
        /**
         * Character code.
         *
         * @var {NonNullable<Code>} code
         */
        let code: NonNullable<Code> = value[index]!.codePointAt(0)!

        /**
         * Difference between next column and current column.
         *
         * @var {number} k
         */
        let k: number = 1

        switch (true) {
          case code === codes.cr:
            if (value[index + 1]?.codePointAt(0) === codes.lf) {
              chunks.push(codes.crlf)
              k++
            } else {
              chunks.push(codes.vcr)
            }

            column = 1
            break
          case code === codes.ht:
            /**
             * Next column.
             *
             * @const {number} n
             */
            const n: number = Math.ceil(column / tabSize) * tabSize

            chunks.push(codes.vht)
            while (column++ < n) chunks.push(codes.vs)

            break
          case code === codes.lf:
            chunks.push(codes.vlf)
            column = 1
            break
          default:
            chunks.push(code)
            column++
            break
        }

        index += k
      }
    }

    return end && chunks.push(codes.eof), chunks
  }
}

export default preprocess
