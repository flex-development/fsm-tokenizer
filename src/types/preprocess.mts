/**
 * @file Type Aliases - Preprocess
 * @module vfile-tokenizer/types/Preprocess
 */

import type {
  Code,
  Encoding,
  FileLike,
  Value
} from '@flex-development/vfile-tokenizer'

/**
 * Turn `value` into character code chunks.
 *
 * @see {@linkcode Code}
 * @see {@linkcode Encoding}
 * @see {@linkcode FileLike}
 * @see {@linkcode Value}
 *
 * @this {void}
 *
 * @param {FileLike | Value | null | undefined} value
 *  The value to preprocess
 * @param {Encoding | null | undefined} [encoding]
 *  The character encoding to use when `value`
 *  or its contents is {@linkcode Uint8Array}
 * @param {boolean | null | undefined} [end]
 *  Whether the end of stream has been reached
 * @return {Code[]}
 *  Character code chunks
 */
type Preprocess = (
  this: void,
  value: FileLike | Value | null | undefined,
  encoding?: Encoding | null | undefined,
  end?: boolean | null | undefined
) => Code[]

export type { Preprocess as default }
