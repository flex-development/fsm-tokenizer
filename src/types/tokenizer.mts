/**
 * @file Type Aliases - Tokenizer
 * @module fsm-tokenizer/types/Tokenizer
 */

import type {
  Effects,
  State,
  TokenizeContext
} from '@flex-development/fsm-tokenizer'

/**
 * Set up a state machine to handle character codes streaming in.
 *
 * @see {@linkcode Effects}
 * @see {@linkcode State}
 * @see {@linkcode TokenizeContext}
 *
 * @this {TokenizeContext}
 *
 * @param {Effects} effects
 *  Context object to transition state machine
 * @param {State} ok
 *  Successful tokenization state
 * @param {State} nok
 *  Failed tokenization state
 * @return {State}
 *  Initial state
 */
type Tokenizer = (
  this: TokenizeContext,
  effects: Effects,
  ok: State,
  nok: State
) => State

export type { Tokenizer as default }
