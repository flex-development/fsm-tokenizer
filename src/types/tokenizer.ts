/**
 * @file Type Aliases - Tokenizer
 * @module vfile-lexer/types/Tokenizer
 */

import type { Effects, TokenizeContext } from '#src/interfaces'
import type State from './state'

/**
 * Set up a state machine to handle character codes streaming in.
 *
 * @see {@linkcode Effects}
 * @see {@linkcode State}
 * @see {@linkcode TokenizeContext}
 *
 * @this {TokenizeContext}
 *
 * @param {Effects} effects - Context object to transition state machine
 * @param {State} ok - Successful tokenization state
 * @param {State} nok - Failed tokenization state
 * @return {State} Initial state
 */
type Tokenizer = (
  this: TokenizeContext,
  effects: Effects,
  ok: State,
  nok: State
) => State

export type { Tokenizer as default }
