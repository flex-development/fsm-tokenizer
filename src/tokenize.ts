/**
 * @file tokenize
 * @module vfile-lexer/tokenize
 */

import type {
  Encoding,
  Event,
  FileLike,
  TokenizeOptions,
  Value
} from '#src/types'
import Lexer from './lexer'
import preprocess from './preprocess'

/**
 * Tokenize `value`.
 *
 * @see {@linkcode Encoding}
 * @see {@linkcode Event}
 * @see {@linkcode FileLike}
 * @see {@linkcode TokenizeOptions}
 * @see {@linkcode Value}
 *
 * @param {FileLike | Value | null | undefined} value - Value to tokenize
 * @param {Encoding | null | undefined} [encoding] - Character encoding to use
 * when `value` or its contents is {@linkcode Uint8Array}
 * @param {TokenizeOptions | null | undefined} [options] - Configuration options
 * @return {Event[]} List of events
 */
function tokenize(
  value: FileLike | Value | null | undefined,
  encoding?: Encoding | null | undefined,
  options?: TokenizeOptions | null | undefined
): Event[]

/**
 * Tokenize `value`.
 *
 * @see {@linkcode FileLike}
 * @see {@linkcode Event}
 * @see {@linkcode TokenizeOptions}
 * @see {@linkcode Value}
 *
 * @param {FileLike | Value | null | undefined} value - Value to tokenize
 * @param {TokenizeOptions | null | undefined} [options] - Tokenize options
 * @return {Event[]} List of events
 */
function tokenize(
  value: FileLike | Value | null | undefined,
  options?: TokenizeOptions | null | undefined
): Event[]

/**
 * Tokenize `value`.
 *
 * @see {@linkcode FileLike}
 * @see {@linkcode Event}
 * @see {@linkcode TokenizeOptions}
 * @see {@linkcode Value}
 *
 * @param {FileLike | Value | null | undefined} value - Value to tokenize
 * @param {Encoding | TokenizeOptions | null | undefined} [encoding] - Character
 * encoding to use when `value` or its contents is {@linkcode Uint8Array}, or
 * configuration options
 * @param {TokenizeOptions | null | undefined} [options] - Configuration options
 * @return {Event[]} List of events
 */
function tokenize(
  value: FileLike | Value | null | undefined,
  encoding?: Encoding | TokenizeOptions | null | undefined,
  options?: TokenizeOptions | null | undefined
): Event[] {
  if (typeof encoding === 'object' && encoding) {
    options = encoding
    encoding = undefined
  }

  return new Lexer(options).write(preprocess(options)(value, encoding, true))
}

export default tokenize
