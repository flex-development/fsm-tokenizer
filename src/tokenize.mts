/**
 * @file tokenize
 * @module vfile-tokenizer/tokenize
 */

import tokenizer from '#create-tokenizer'
import preprocess from '#preprocess'
import type {
  Event,
  FileLike,
  TokenizeOptions,
  Value
} from '@flex-development/vfile-tokenizer'

/**
 * Tokenize `value`.
 *
 * @see {@linkcode Event}
 * @see {@linkcode FileLike}
 * @see {@linkcode TokenizeOptions}
 * @see {@linkcode Value}
 *
 * @this {void}
 *
 * @param {FileLike | Value | null | undefined} value
 *  File or value
 * @param {TokenizeOptions} options
 *  Configuration options
 * @return {Event[]}
 *  List of events
 */
function tokenize(
  this: void,
  value: FileLike | Value | null | undefined,
  options: TokenizeOptions
): Event[] {
  return tokenizer(options).write(preprocess(options)(
    value,
    options.encoding,
    true
  ))
}

export default tokenize
