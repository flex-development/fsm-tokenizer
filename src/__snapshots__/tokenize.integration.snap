// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`integration:tokenize > default constructs > should tokenize empty file 1`] = `
tokens[2]
├─0 sof (1:1-1:1, 0-0)
└─1 eof (1:1-1:1, 0-0)
`;

exports[`integration:tokenize > default constructs > should tokenize non-empty file 1`] = `
tokens[2]
├─0 sof (1:1-1:1, 0-0)
└─1 eof (22:1-22:1, 304-304)
`;

exports[`integration:tokenize > user constructs > non-empty file > sample 0 1`] = `
tokens[3]
├─0 sof (1:1-1:1, 0-0)
├─1 inlineTag (1:1-1:17, 0-16)
└─2 eof (2:1-2:1, 17-17)
`;

exports[`integration:tokenize > user constructs > non-empty file > sample 1 1`] = `
tokens[13]
├─0  sof (1:1-1:1, 0-0)
├─1  punctuator "=" (1:10-1:11, 9-10)
├─2  string "\\"hello 👋\\"" (1:12-1:21, 11-20)
├─3  punctuator ";" (1:21-1:22, 20-21)
├─4  punctuator "." (2:8-2:9, 29-30)
├─5  punctuator "(" (2:12-2:13, 33-34)
├─6  punctuator "\\\\" (2:13-2:14, 34-35)
├─7  punctuator "\\\\" (2:19-2:20, 40-41)
├─8  punctuator ")" (2:25-2:26, 46-47)
├─9  punctuator ";" (2:26-2:27, 47-48)
├─10 punctuator "/" (2:28-2:29, 49-50)
├─11 punctuator "/" (2:29-2:30, 50-51)
└─12 eof (3:1-3:1, 60-60)
`;

exports[`integration:tokenize > user constructs > non-empty file > sample 2 1`] = `
tokens[8]
├─0 sof (1:1-1:1, 0-0)
├─1 string "'😍'" (1:1-1:4, 0-3)
├─2 string "\\"👍\\"" (2:1-2:4, 4-7)
├─3 punctuator "\\\\" (3:1-3:2, 8-9)
├─4 punctuator "'" (3:2-3:3, 9-10)
├─5 punctuator "\\\\" (3:4-3:5, 11-12)
├─6 punctuator "'" (3:5-3:6, 12-13)
└─7 eof (4:1-4:1, 14-14)
`;

exports[`integration:tokenize > user constructs > non-empty file > sample 3 1`] = `
tokens[42]
├─0  sof (1:1-1:1, 0-0)
├─1  number "0" (1:1-1:2, 0-1)
├─2  whitespace "\\n" (1:2-2:1, 1-2)
├─3  bigint "0n" (2:1-2:3, 2-4)
├─4  whitespace "\\n" (2:3-3:1, 4-5)
├─5  number "1" (3:1-3:2, 5-6)
├─6  whitespace "\\n" (3:2-4:1, 6-7)
├─7  bigint "1n" (4:1-4:3, 7-9)
├─8  whitespace "\\n" (4:3-5:1, 9-10)
├─9  number "2" (5:1-5:2, 10-11)
├─10 whitespace "\\n" (5:2-6:1, 11-12)
├─11 bigint "2n" (6:1-6:3, 12-14)
├─12 whitespace "\\n" (6:3-7:1, 14-15)
├─13 number "3" (7:1-7:2, 15-16)
├─14 whitespace "\\n" (7:2-8:1, 16-17)
├─15 bigint "3n" (8:1-8:3, 17-19)
├─16 whitespace "\\n" (8:3-9:1, 19-20)
├─17 number "4" (9:1-9:2, 20-21)
├─18 whitespace "\\n" (9:2-10:1, 21-22)
├─19 bigint "4n" (10:1-10:3, 22-24)
├─20 whitespace "\\n" (10:3-11:1, 24-25)
├─21 number "5" (11:1-11:2, 25-26)
├─22 whitespace "\\n" (11:2-12:1, 26-27)
├─23 bigint "5n" (12:1-12:3, 27-29)
├─24 whitespace "\\n" (12:3-13:1, 29-30)
├─25 number "6" (13:1-13:2, 30-31)
├─26 whitespace "\\n" (13:2-14:1, 31-32)
├─27 bigint "6n" (14:1-14:3, 32-34)
├─28 whitespace "\\n" (14:3-15:1, 34-35)
├─29 number "7" (15:1-15:2, 35-36)
├─30 whitespace "\\n" (15:2-16:1, 36-37)
├─31 bigint "7n" (16:1-16:3, 37-39)
├─32 whitespace "\\n" (16:3-17:1, 39-40)
├─33 number "8" (17:1-17:2, 40-41)
├─34 whitespace "\\n" (17:2-18:1, 41-42)
├─35 bigint "8n" (18:1-18:3, 42-44)
├─36 whitespace "\\n" (18:3-19:1, 44-45)
├─37 number "9" (19:1-19:2, 45-46)
├─38 whitespace "\\n" (19:2-20:1, 46-47)
├─39 bigint "9n" (20:1-20:3, 47-49)
├─40 whitespace "\\n" (20:3-21:1, 49-50)
└─41 eof (21:1-21:1, 50-50)
`;

exports[`integration:tokenize > user constructs > should tokenize empty file 1`] = `
tokens[2]
├─0 sof (1:1-1:1, 0-0)
└─1 eof (1:1-1:1, 0-0)
`;
